### PR TITLE
#168064097 Fix tag content and response message

### PR DESCRIPTION
**Pull request by Nennyfills**
#### What does this PR do?
This pull request fixes the bug on tag content, response message and pagination count on following and search article endpoint.

#### Description of Task to be completed
- Rearrange tag content on search articles endpoint
- Correct response message on followings endpoint
- Add count on getting all followers and followings endpoint

#### How should this be manually tested?
- Clone this repo
Using an HTTP client test the following endpoints
- ``` GET /api/v1/:articles?```
- ``` patch /api/v1/users/profile/:userId/following```
- ```GET /api/v1/users/profile/:userId/followers```
- ```GET /api/v1/users/profile/:userId/followings```

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#168064097](https://pivotaltracker.com/story/show/168064097)

#### Screenshots (if appropriate)
Screen Shot 2019-08-22 at 3 43 54 PM
Screen Shot 2019-08-22 at 4 05 25 PM
Screen Shot 2019-08-22 at 4 05 38 PM
Screen Shot 2019-08-22 at 4 21 16 PM
Screen Shot 2019-08-22 at 7 08 56 PM
Screen Shot 2019-08-23 at 11 48 58 AM
Screen Shot 2019-08-23 at 11 58 15 AM

#### Questions:
N/A

### CODE REVIEW

**Comment by iverenshaguy** [comment]
Please flatten the response of `/followees`. Please also add a screenshot of what all these responses look like with empty data

File: docs/swagger.json [hunk 1]
```diff
@@ -885,7 +885,7 @@
       }
     },
     "/profile/:userId/following": {
-      "post": {
+      "patch": {
         "description": "follower a user",
```
Comment (iverenshaguy) [inline docs/swagger.json]:
```suggestion
        "description": "follow a user",
```

File: docs/swagger.json [hunk 2]
```diff
@@ -885,7 +885,7 @@
       }
     },
     "/profile/:userId/following": {
-      "post": {
+      "patch": {
```
Comment (iverenshaguy) [inline docs/swagger.json:915]: Why was this changed?

File: src/helpers/genericArticleHelper.js [hunk 3]
```diff
@@ -5,22 +5,22 @@ const { Op } = Sequelize;
 const { Users, Categories, Articles } = models;
```
Comment (iverenshaguy) [inline src/helpers/genericArticleHelper.js]: Please rename this file to something like `articleHelper.js`

File: src/helpers/genericArticleHelper.js [hunk 4]
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
   if (!tag) delete queryParams.query['$tag.category$'];
 
   // eslint-disable-next-line no-nested-ternary
-  const filterQuery = (keyword) ? { ...queryParams.keyword } : author ? { ...queryParams.author }
-    : { ...queryParams.query };
+  const filterQuery = keyword
```
Comment (iverenshaguy) [inline src/helpers/genericArticleHelper.js]: Please refactor this piece of code to something more readable. We want to avoid nested ternary operations at all times

File: src/helpers/genericArticleHelper.js [hunk 5]
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
...
-    } : undefined
+    include: [
+      { model: Users, as: 'author' },
+      { model: Articles, as: 'article' },
+      { model: Categories, as: 'tag' }
+    ],
+    where:
+      keyword || title || tag || author
+        ? {
+          ...filterQuery
```
Comment (iverenshaguy) [inline src/helpers/genericArticleHelper.js]:
```suggestion
          { ...filterQuery  }
```

File: src/helpers/genericArticleHelper.js [hunk 6]
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
...
+        }
+        : undefined
```
Comment (iverenshaguy) [inline src/helpers/genericArticleHelper.js]: If line 60 is not true, where would be undefined anyways so please change it to something like
```suggestion
const hasFilterQuery = keyword || title || tag || author;
...
where: hasFilterQuery && { ...filterQuery  }
...
```

File: src/routes/user.js [hunk 7]
```diff
@@ -90,7 +90,7 @@ router.get('/profile/:id', validator(profileViewSchema), asyncWrapper(verifyToke
  * @swagger
  *
  * /profile/:userId/following:
- *   post:
+ *   patch:
```
Comment (iverenshaguy) [inline src/routes/user.js:100]: Why was this changed?

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Please review my comments and write tests for uncovered code

File: src/routes/user.js (hunk 7 above)
Comment (Nennyfills) [reply src/routes/user.js:100]: I changed that part because the route is a patch route, not post route.

File: docs/swagger.json (hunk 2 above)
Comment (Nennyfills) [reply docs/swagger.json:915]: I changed it because is a patch route and not a post route.

File: docs/swagger.json (hunk 1 above)
Comment (Nennyfills) [reply docs/swagger.json]: Done.

File: src/helpers/genericArticleHelper.js (hunk 3 above)
Comment (Nennyfills) [reply src/helpers/genericArticleHelper.js]: Done.

File: src/helpers/genericArticleHelper.js (hunk 4 above)
Comment (Nennyfills) [reply src/helpers/genericArticleHelper.js]: Done.

File: src/helpers/genericArticleHelper.js (hunk 5 above)
Comment (Nennyfills) [reply src/helpers/genericArticleHelper.js]: Done.

File: src/helpers/genericArticleHelper.js (hunk 6 above)
Comment (Nennyfills) [reply src/helpers/genericArticleHelper.js]: Done.

File: src/routes/user.js (hunk 7 above)
Comment (Nennyfills) [reply src/routes/user.js:100]: I changed it because is a patch route and not a post route.

File: docs/swagger.json (hunk 2 above)
Comment (iverenshaguy) [reply docs/swagger.json:915]: What makes it a patch route?

Comment (iverenshaguy) [reply docs/swagger.json:915]: I don't really understand. How did we come to that conclusion?

**Comment by iverenshaguy** [comment]
The response structure that has `following: []` should be `data`. We want to ensure consistency in our codebase

File: src/controllers/user.js [hunk 8]
```diff
@@ -138,9 +138,14 @@ export default {
     return response.status(200).json({
       status: 'success',
       message: 'Request successful',
-      data,
-      count,
-      page
+      followers: data.map((item) => {
```
Comment (iverenshaguy) [inline src/controllers/user.js]: followers should be data ie.e 
```suggestion
      data: whatever.map((item) => {
```

File: src/controllers/user.js [hunk 9]
```diff
@@ -168,10 +173,16 @@ export default {
 
     return response.status(200).json({
       status: 'success',
-      data,
+      message: 'Request successful',
+      followings: data.map((item) => {
```
Comment (iverenshaguy) [inline src/controllers/user.js]: Same feedback
```suggestion
      data: whatever.map((item) => {
```

**Review by iverenshaguy — CHANGES_REQUESTED** [review]

File: src/helpers/articleHelper.js [hunk 10]
```diff
@@ -5,17 +5,17 @@ const { Op } = Sequelize;
 const { Users, Categories, Articles } = models;
 
 /**
-   * filter for search article
```
Comment (iverenshaguy) [inline src/helpers/articleHelper.js]: So I saw that we have another file called `article` in this helpers folder. I don't really understand why the methods here were not added to that file (I'd rather they were). Is there a reason why we decided to create another file called `genericArticleHelper`? Also, I need to understand why you feel the need to use the word `generic` in the codebase. I have seen you describe functions as `generic ...` in your JSDocs. Is there a specific reason for this?

File: src/controllers/user.js (hunk 8 above)
Comment (Nennyfills) [reply src/controllers/user.js]: Done.

File: src/controllers/user.js (hunk 9 above)
Comment (Nennyfills) [reply src/controllers/user.js]: Done.

File: src/helpers/articleHelper.js (hunk 10 above)
Comment (Nennyfills) [reply src/helpers/articleHelper.js]: Before the search article PR was merged the other article file wasn't there. I called it ```genericArticleHelper``` because the sequelize operators used in the file are not limited. it can be reused anywhere to find any article or author. The function can be reused on any route to validate an empty body.

Comment (iverenshaguy) [reply src/helpers/articleHelper.js]: Okay. Please move all methods in this file to the `./helpers/article.js` file. Also, I don't think we need to use generic. Helpers can generally be used anywhere so there's no need to specify. Thanks

**Comment by iverenshaguy** [comment]
Please rebase and fix conflicts. Also, I have still not understood why the route was changed from post to patch.
